File: app/src/main/java/xyz/zpayh/myadapter/AutoLoadMoreActivity.java
package xyz.zpayh.myadapter;

import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

import xyz.zpayh.adapter.OnLoadMoreListener;
import xyz.zpayh.myadapter.adapter.AutoLoadAdapter;
import xyz.zpayh.myadapter.data.ImageCard;

public class AutoLoadMoreActivity extends AppCompatActivity implements View.OnClickListener{

    public static final int LOAD_ADD = 0;
    public static final int LOAD_FAILED = 1;
    public static final int LOAD_COMPLETED = 2;

    private int mState = LOAD_ADD;

    private AutoLoadAdapter mAdapter;

    private List<ImageCard> data;

    private String mTitles[] = {"Adult","Easter Eggs","Girl", "Sunset"};
    private int mImageResId[] = {R.drawable.adult, R.drawable.easter_eggs, R.drawable.girl, R.drawable.sunset};

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.act_auto_load_more);

        final RecyclerView recyclerView = (RecyclerView) findViewById(R.id.list);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        mAdapter = new AutoLoadAdapter(this);
        recyclerView.setAdapter(mAdapter);
        final SwipeRefreshLayout refreshLayout = (SwipeRefreshLayout) findViewById(R.id.refresh);

        mAdapter.addHeadLayout(R.layout.item_head);
        mAdapter.addFootLayout(R.layout.item_foot2);

        // 模拟数据
        data = new ArrayList<>();

        for (int i = 0; i < mTitles.length; i++) {
            ImageCard card = new ImageCard(mImageResId[i],mTitles[i]);
            data.add(card);
        }

        refreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                //模拟刷新
                recyclerView.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        refreshLayout.setRefreshing(false);
                        mAdapter.setData(data);
                    }
                },500);
            }
        });

        //必须设置事件监听与开启auto
        mAdapter.openAutoLoadMore(true);
        mAdapter.setOnLoadMoreListener(new OnLoadMoreListener() {
            @Override
            public void onLoadMore() {
                Log.d("Sherlock","加载更多");
                //模拟加载更多
                recyclerView.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        if (mState == LOAD_ADD){
                            mAdapter.addData(data);
                        }else if (mState == LOAD_COMPLETED){
                            mAdapter.loadCompleted();
                        }else if (mState == LOAD_FAILED){
                            mAdapter.loadFailed();
                        }
                    }
                },800);
            }
        });

        findViewById(R.id.action_add).setOnClickListener(this);
        findViewById(R.id.action_failed).setOnClickListener(this);
        findViewById(R.id.action_completed).setOnClickListener(this);
        findViewById(R.id.action_empty).setOnClickListener(this);
        findViewById(R.id.action_close).setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()){
            case R.id.action_add:
                setTitle("加载更多");
                mAdapter.openAutoLoadMore(true);
                mAdapter.setData(data);
                mState = LOAD_ADD;
                break;
            case R.id.action_failed:
                setTitle("加载更多失败");
                mAdapter.openAutoLoadMore(true);
                mAdapter.setData(data);
                mState = LOAD_FAILED;
                break;
            case R.id.action_completed:
                setTitle("没有更多数据");
                mAdapter.openAutoLoadMore(true);
                mAdapter.setData(data);
                mState = LOAD_COMPLETED;
                break;
            case R.id.action_empty:
                setTitle("没有数据");
                mAdapter.setData(null);
                break;
            case R.id.action_close:
                mAdapter.openAutoLoadMore(false);
                break;
        }
    }
}

File: sample-kotlin/src/main/java/cn/sherlockzp/sample/AutoLoadMoreActivity.kt
package cn.sherlockzp.sample

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import cn.sherlockzp.adapter.BaseAdapter
import cn.sherlockzp.adapter.BaseExpandableAdapter
import cn.sherlockzp.adapter.BaseViewHolder
import cn.sherlockzp.adapter.LayoutConfig
import cn.sherlockzp.sample.model.Card
import cn.sherlockzp.sample.model.ImageCard
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import kotlinx.android.synthetic.main.activity_auto_load_more.*

class AutoLoadMoreActivity : AppCompatActivity() , View.OnClickListener{

    private val LOAD_ADD = 0
    private val LOAD_FAILED = 1
    private val LOAD_COMPLETED = 2

    private var state = LOAD_ADD

    private var showEmpty = true

    private val width by lazy {
        resources.displayMetrics.widthPixels
    }

    private val titles = arrayOf("Adult","Easter Eggs","Girl", "Sunset")
    private val imageResIds = arrayOf(R.drawable.adult, R.drawable.easter_eggs, R.drawable.girl, R.drawable.sunset)

    private val data = ArrayList<ImageCard>()
    private val adapter = object : BaseAdapter<ImageCard>() {
        override fun getLayoutRes(index: Int) = R.layout.item_image_card

        override fun convert(holder: BaseViewHolder, data: ImageCard, index: Int) {
            holder.setText(R.id.tv_image_title, data.title)
                    .setImage(R.id.iv_bg){
                        Glide.with(it.context)
                                .load(data.imageResId)
                                .apply(RequestOptions.fitCenterTransform())
                                .apply(RequestOptions.overrideOf(width))
                                .into(it)
                    }
        }

        override fun convertHead(holder: BaseViewHolder, headLayout: LayoutConfig) {
            holder.setText(R.id.tv_head, "这是加载更多")
        }
    }
    /*
    private val data = ArrayList<Card>()
    private val adapter = object : BaseExpandableAdapter() {
        override fun convertHead(holder: BaseViewHolder, headLayout: LayoutConfig) {
            holder.setText(R.id.tv_head, "这是加载更多")
        }
    }*/

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_auto_load_more)

        initView()
        initData()
    }

    private fun initView(){
        rv_list_load_more.layoutManager = LinearLayoutManager(this)
        rv_list_load_more.adapter = adapter

        with(adapter) {
            alwaysShowFoot = true
            alwaysShowHead = true
            addHeadLayout(R.layout.item_head)
            addFootLayout(R.layout.item_foot2)

            // 必须同时设置事件监听与开启auto
            openAutoLoadMore()
            setOnLoadMoreListener {
                // 模拟加载更多
                rv_list_load_more.postDelayed({
                    when(state) {
                        LOAD_ADD -> adapter.addData(this@AutoLoadMoreActivity.data)
                        LOAD_COMPLETED -> adapter.loadCompleted()
                        LOAD_FAILED -> adapter.loadFailed()
                    }
                }, 800)
            }
        }

        sr_refresh.setOnRefreshListener {
            // 模拟刷新
            rv_list_load_more.postDelayed({
                sr_refresh.isRefreshing = false
                adapter.setData(data)
            }, 500)
        }

        cb_show_head.isChecked = true
        cb_show_head.setOnCheckedChangeListener{
            _,isChecked -> adapter.alwaysShowHead = isChecked
        }
        cb_show_foot.isChecked = true
        cb_show_foot.setOnCheckedChangeListener{
            _,isChecked -> adapter.alwaysShowFoot = isChecked
        }

        action_add.setOnClickListener(this)
        action_failed.setOnClickListener(this)
        action_completed.setOnClickListener(this)
        action_empty.setOnClickListener(this)
        action_close.setOnClickListener(this)
    }

    private fun initData() {
        for (i in titles.indices) {
            data.add(ImageCard(imageResIds[i], titles[i]))
            //data.add(Card(width, imageResIds[i], titles[i]))
        }
        adapter.setData(data)
    }

    override fun onClick(v: View?) {
        when(v) {
            action_add -> {
                title = "加载更多"
                adapter.openAutoLoadMore()
                adapter.setData(data)
                state = LOAD_ADD
            }
            action_failed -> {
                title = "加载更多失败"
                adapter.openAutoLoadMore()
                adapter.setData(data)
                state = LOAD_FAILED
            }
            action_completed -> {
                title = "没有更多数据"
                adapter.openAutoLoadMore()
                adapter.setData(data)
                state = LOAD_COMPLETED
            }
            action_empty -> {
                if (showEmpty) {
                    title = "没有数据"
                    adapter.setData(null)
                    showEmpty = false
                    action_empty.text = "异常数据"
                } else {
                    title = "数据异常"
                    adapter.showErrorView()
                    showEmpty = true
                    action_empty.text = "空数据"
                }
            }
            action_close -> {
                adapter.openAutoLoadMore(false)
            }
        }
    }
}
